knowledge-base/how-to-inject-custom-dependency-in-reports-controller-of-dot-net-core-application.md

```diff
@@ -0,0 +1,128 @@
+---
+title: Inject custom dependency in the ReportsController of a .NET Core application
+description: How to inject custom dependency in the ReportsController of a .NET Core application
+type: how-to
+page_title: Dependency injection in the ReportsController of a .NET Core application
+slug: how-to-inject-custom-dependency-in-reports-controller-of-dot-net-core-application
+position: 
+tags: 
+ticketid: 1418590
+res_type: kb
+---
+
+## Environment
+<table>
+    <tbody>
+	    <tr>
+	    	<td>Product</td>
+	    	<td>Progress® Telerik® Reporting</td>
+	    </tr>
+	    <tr>
+	    	<td>Project Type</td>
+	    	<td>ASP.NET Core</td>
+	    </tr>
+    </tbody>
+</table>
+
+
+## Description
+ASP.NET Core provides built-in functionality for dependency injection. By default, the ReportsController, where the Telerik Reporting REST Service gets configured receives only an object of custom configuration class (e.g. _ConfigurationService_ in our examples), where the [IHostingEnvironment](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.hosting.ihostingenvironment?view=aspnetcore-2.2) and [IConfiguration](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.configuration.iconfiguration?view=aspnetcore-2.2) are passed.  
+If you need to pass additional dependency, you may do it as explained below.
+
+## Solution
+
+Let's demonstrate the approach with an example - how to provide a data source through inversion of control in the report dynamically.
+The Telerik Reporting uses [System.Reflection](https://docs.microsoft.com/en-us/dotnet/api/system.reflection?view=netcore-2.2) to create instances of the custom objects at run-time. Therefore, they need public parameterless constructor in order to be successfully created. Hence, the ObjectDataSource cannot use a provider (e.g. **ProductRepository**) where a dependency (e.g. the database context) is injected, for example like :
+
+```CSharp
+public class ProductRepository : IProductRepository
+{
+    private readonly ApplicationDbContext context;
+
+    public ProductRepository(ApplicationDbContext context)
+    {
+        this.context = context;
+    }
+
+    public IEnumerable<Product> GetProducts()
+    {
+        // return the data here, e.g.
+        return this.context.Product.Take(20).ToList();
+    }
+}
+```
+  
+In this case you may use [Custom Resolver](../telerik-reporting-rest-custom-report-resolver) (e.g. **CustomReportResolver**) for resolving the TRDP reports and add the DataSource at run time after instantiating the report. For the purpose it will be necessary to :  
+
+1) Change the **Configure** method of the **Startup** class to pass the **IProductRepository** with the **ReportsController**, for example like :  
```

Comment:
please use this cheat sheet https://wiki.progress.com/display/RST/KB+Markdown+Cheat-sheet to see how the numeration followed by a code snipper will work when the documentation is released.

//new row
1. First step
//new row
    //tab
     ```
     code-snippet here
     ```
//no empty row after the code snippet
2. Second step